services/comments.js

```diff
@@ -49,70 +53,77 @@ module.exports = class CommentsService {
-        assigned_by,
-        created_at: new Date()
-      }}
-    };
-    return CommentModel.update(filter, update)
-      .then(({nModified}) => {
-        switch (nModified) {
-        case 0:
-
-          // either the tag was already there, or the comment doesn't exist with that id...
-          throw new Error('Could not add tag to comment. Either the comment doesn\'t exist or the tag is already present.');
-        case 1:
-
-            // tag added
-          return;
-        default:
-
-          // this should never happen because no multi parameter and unique index on id
+  static addTag(id, name, added_by) {
+
+    // Check that the tag is allowed by the system OR in our setting.tags.
+    return SettingsService.retrieve()
+    .then((settings) => {
+
+      UsersService.findById(added_by)
+      .then((user) => {
+
+        // Moderators or ADMIN can add any tag automatically.
+        if (user != null && (user.hasRoles('ADMIN') || user.hasRoles('MODERATOR'))) {
+          SettingModel.findOneAndUpdate({id: settings.id}, {
+            $push: {
+              tags: {
+                id: name,
+                models: ['COMMENTS']
+              }
+            }
+          });
+        }
+        else if (!ALLOWED_TAGS.includes(name) || settings.tags.findIndex((t) => {return t.id === name & t.models.include('COMMENTS');}) === -1) {
+          return Promise.reject(errors.ErrTagNotAllowed);
         }
       });
+
+      return CommentModel.findOneAndUpdate({id, 'tags.id': {$ne: name}}, {
+        $push: {
+          tags: {
+            id: name,
+            added_by: added_by
+          }
+        },
+      })
+        .then(({nModified}) => {
+          switch (nModified) {
+          case 0:
+            return Promise.reject(errors.ErrNoCommentFound);
+          case 1:
+            return;
+          default:
+          }
+        });
+    });
   }
 
   /**
    * Removes a tag from a comment
    * @throws if the tag is not on the comment
    * @param {String} id the id of the comment to tag
-   * @param {String} name the name of the tag to add
+   * @param {String} tag_id the id of the tag to remove
    */
-  static removeTag(id, name) {
-    const filter = {
-      id,
-      'tags.name': name,
-    };
-    const update = {$pull: {tags: {name}}};
-    return CommentModel.update(filter, update)
-      .then(({nModified}) => {
-        switch (nModified) {
-        case 0:
-          throw new Error('Could not remove tag from comment. Either the comment doesn\'t exist or the tag is not present');
-        case 1:
-
-            // tag removed
-          return;
-        default:
-
-          // this should never happen because no multi parameter and unique index on id
+  static removeTag(id, tag_id) {
+    return CommentModel.findOneAndUpdate({id, 'tags.id': tag_id}, {
+      $pull: {
+        tags: {
+          id: tag_id
         }
-      });
+      }
+    }
+    )
+    .then(({nModified}) => {
+      switch(nModified) {
```

Comment:
If you are just performing an update, then just use the update instead of the `findOneAndUpdate`. Also, there is a `nMatched` that should be used to check to see if the query found something.